File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int i = 0  ; 
        int j = 0 ; 
        int ans = 0 ; 

        int[] arr = new int[26] ; 

        while ( i < s.length() ) {
            char ch = s.charAt(i) ; 

            arr[ch-'a']++ ; 
            while(arr[ch-'a'] > 1 ) {
                char prev = s.charAt(j) ; 
                arr[prev-'a']-- ; 
                j++ ; 
            }
            ans = Math.max(ans , i - j + 1 ) ; 
            i++ ; 
        }
        return ans ; 
    }
}

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
class Solution(object):
    def lengthOfLongestSubstring(self, s):
        """
        :type s: str
        :rtype: int
        """
        i = 0 
        j = 0 
        max_len = 0 
        dict1 = {}

        while j < len(s) :
            chars = s[j] 
            dict1[chars] = dict1.get(chars,0) + 1 

            while dict1[chars] > 1 :
                ch = s[i] 
                dict1[ch] = dict1[ch] - 1 
                i += 1 

            max_len = max(max_len , j - i + 1 )   
            j += 1  

        return max_len           
